### PR TITLE
[breaking] large refactoring of SCS.jl

### LINKED ISSUE #217: SCS abuses convert
Stuff like this
https://github.com/jump-dev/SCS.jl/blob/500e75d8f3a3467b8ae2814a49e41aa05c31439e/src/types.jl#L30-L52
makes me nervous.

We should thoroughly audit the use of cconvert in SCS.jl. We're passing a lot of `Ref{SCSData}` to `ccall`, when best practice is to implement `unsafe_convert` and `cconvert`. This should limit the need for `GC.@preserve` etc.

### PR DESCRIPTION
**Pull request by odow**
This PR is a large refactoring of the C code in SCS.jl and fixes #217. It doesn't touch the MOI wrapper.

It removes all exported functions except `SCS_solve`. We don't want users touching the other functions because there's too much room for GC errors.

There is now a single `_SCSDataWrapper` struct which prevents stuff from being GC'd, and there is now a single `_unsafe_scs_solve` function in which the pointer stuff takes place. 

It also explicitly refactors the linear solvers into three files, and removes all the nasty `@eval` code.

I've marked this as v0.8.0, but we should hold off tagging a new release until we update MOI as well.

### CODE REVIEW

**Review by kalmarek — CHANGES_REQUESTED** [review]
I took a short look at the changes and I'm fine with those;

I left one small comment on `_to_sparse`.

A few additional points:
 * in scs-3.0.0 `scsint_t` would become a `ccall`
 * when we drop julia-1.0 (hopefully soon) and jump to the new LTS the ccalls could be made fully dynamic (aka non-constant function specifications) and then we could merge all separate solvers code again.

**Comment by kalmarek** [comment]
> It removes all exported functions except `SCS_solve`. We don't want users touching the other functions because there's too much room for GC errors.

I always forgot to remove those exports ;)

> It also explicitly refactors the linear solvers into three files, and removes all the nasty @eval code.

I like `@eval`s ;) and when we drop old julias and move to dynamically generated ccalls these three files could be merged again ;)

File: src/MOI_wrapper/matrix.jl [hunk 1]
```diff
@@ -12,6 +12,10 @@ mutable struct SparseMatrixCSRtoCSC{T}
     end
 end
 
+function _to_sparse(::Type{T}, A::SparseMatrixCSRtoCSC{T}) where {T}
+    return A.nzval, A.rowval, A.colptr
```
[the remark this answers is not in the document]
Comment (kalmarek) [reply src/MOI_wrapper/matrix.jl:16]: I've seen similar SparseMatrixCSRtoCSC in MOI: https://github.com/jump-dev/MatrixOptInterface.jl/blob/master/src/sparse_matrix.jl

Do we even need our own copy?

[the remark this answers is not in the document]
Comment (odow) [reply src/MOI_wrapper/matrix.jl:16]: It will be gone in MOI 0.10.

**Comment by odow** [comment]
Yeah presumably we will update to SCS 3.0 before releasing SCS.jl 1.0. This PR is more to tidy some things up before we attack the MOI wrapper.

**Comment by odow** [comment]
Merging because I need this to work on the MOI PR. We can fix any issues in a separate PR; we won't be tagging a new release immediately.

**Comment by kalmarek** [comment]
@odow I don't want to hold back your work on MOI, so merge stuff at will; I hope to find more time next week to rework SCS-3.0.0 PR and then dive deep into all the changes

**Pull request merged**